### PR TITLE
Issue #13672: Kill mutation for DetailAstImpl-3

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #13672: Kill mutation for DetailAstImpl-3

https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java

-------

# Mutations 
https://github.com/checkstyle/checkstyle/blob/88486f7ca9795f613d28fa6c2a53beccef521dfc/config/pitest-suppressions/pitest-common-2-suppressions.xml#L30-L46

------------

# Explaination

To add Next sibiling. The code is 
https://github.com/checkstyle/checkstyle/blob/0dba32b6c8c29977d74af675078501fa8052c10f/src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java#L140-L156

Line 150 and 153 is mutated 

This explaination is almost similar to https://github.com/checkstyle/checkstyle/pull/13390#issue-1800473514

lets suppose `if (sibling != null) {` if this statment execute then 
`astImpl.setNextSibling(sibling);` while calling setNextSibling the previous sibling will be set at https://github.com/checkstyle/checkstyle/blob/0dba32b6c8c29977d74af675078501fa8052c10f/src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java#L473-L475 

For 100% surity we can remove 
https://github.com/checkstyle/checkstyle/blob/0dba32b6c8c29977d74af675078501fa8052c10f/src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java#L466-L476

if last statment from this code in which previous sibiling are set
https://github.com/checkstyle/checkstyle/blob/0dba32b6c8c29977d74af675078501fa8052c10f/src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java#L473-L475

 testing puropse that this happening or not we can remove 
 and know if you run my updated test case it will still run know run this test by removing the mutated code it will fail.

--------------

# Regression 

No Diff

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/part3-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/sevntu-check-regression_part_1-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/checks-nonjavadoc-error-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/sevntu-check-regression_part_2-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/test-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/checks-only-javadoc-error-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/part1-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/part5-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/part6-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/part4-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/antlr-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-25-20/part2-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/part3-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/sevntu-check-regression_part_1-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/checks-nonjavadoc-error-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/sevntu-check-regression_part_2-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/test-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/checks-only-javadoc-error-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/part1-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/part5-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/part6-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/part4-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/antlr-report/index.html

https://kevin222004.github.io/reports/com3/2023-09-06-T-16-30-48/part2-report/index.html

### CODE REVIEW

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java [hunk 1]
```diff
@@ -138,21 +138,15 @@ else if (parent != null) {
      * @param ast DetailAST object.
      */
     public void addNextSibling(DetailAST ast) {
-        clearBranchTokenTypes();
-        clearChildCountCache(parent);
-        if (ast != null) {
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:143]: I see that both `clearXXX` methods are called in `setNextSibling`, so you remove them here as redundant. But I don't understand why you remove the null check `if (ast != null)`, could you explain please?

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java [hunk 2]
```diff
@@ -138,21 +138,15 @@ else if (parent != null) {
...
-            // parent is set in setNextSibling
-            final DetailAstImpl sibling = nextSibling;
-            final DetailAstImpl astImpl = (DetailAstImpl) ast;
+        // parent is set in setNextSibling
+        final DetailAstImpl sibling = nextSibling;
+        final DetailAstImpl astImpl = (DetailAstImpl) ast;
 
-            if (sibling != null) {
-                astImpl.setNextSibling(sibling);
-                sibling.previousSibling = astImpl;
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:150]: In the `astImpl.setNextSibling` we assign `sibling.previousSibling = this`, after that we have `sibling.previousSibling = astImpl`. It doesn't look like `this` is the same link as `astImpl`. Why did you remove it?

File: src/test/java/com/puppycrawl/tools/checkstyle/DetailAstImplTest.java [hunk 3]
```diff
@@ -477,22 +477,59 @@ public void testAddNextSibling() {
...
+        final DetailAstImpl newSibling = new DetailAstImpl();
+        final DetailAST newNextSibling = new DetailAstImpl();
+
         parent.setFirstChild(child);
         child.setNextSibling(sibling);
         child.addNextSibling(newSibling);
-
+        newSibling.addNextSibling(newNextSibling);
+
+        assertWithMessage("")
```
Comment (rdiachenko) [inline src/test/java/com/puppycrawl/tools/checkstyle/DetailAstImplTest.java:488]: Please replace empty strings with proper messages.

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java (hunk 1 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:143]: Ok so lets suppose if ast is null. then 
in current code the process stop at `if (ast != null)`. 

But if not then my updated code is 
```
        final DetailAstImpl sibling = nextSibling;
        final DetailAstImpl astImpl = (DetailAstImpl) ast;

        if (sibling != null) {
            astImpl.setNextSibling(sibling);
        }

        setNextSibling(astImpl);
```

ast is assigned to astImpl and it is used at the end of method ` setNextSibling(astImpl);` (This is the only usage of last in whole method)  know in `setNextSibling(astImpl);`  (where atImpl is null and we are passing null to setNextImpl)

```
     public void setNextSibling(DetailAST nextSibling) {
        clearBranchTokenTypes();
        clearChildCountCache(parent);
        this.nextSibling = (DetailAstImpl) nextSibling;
        if (nextSibling != null && parent != null) {
            ((DetailAstImpl) nextSibling).setParent(parent);
        }
        if (nextSibling != null) {
            ((DetailAstImpl) nextSibling).previousSibling = this;
        }
    }
```
here https://github.com/checkstyle/checkstyle/blob/7f6df92a4f2cf2db3c954577ef856594645e459b/src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java#L60 will be assigned to nextSibiling (which is parameter of method).

know to set the parent of this node and the previous sibling of the node in both the if condition it is assigned that `nextSibiling != null` so it will never be the node of the tree (for surety I have updated the test case to check the childcount of parent you can check `testAddNextSibling2`)and not affect code or generate any kind of null pointer exception.

Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:143]: How about this logic:
```
        if (sibling != null) {
            astImpl.setNextSibling(sibling);
        }
```
It is a potential NPE here when `sibling` is not null but `astImpl` is null. Is this a possible outcome?

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java (hunk 2 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:150]: For Explanation, I will take an example. 

Let's suppose,
We have some structure where we have nodes `X, Y, Z, W`  where `X` is the parent node `Y` is the first child and `Z` is a sibling of `Y`
```
 X
 |
  ----Y-----Z

```

Now we have to add the `W` node as nextSibiling of `Y` 
```
 X
 |
  ----Y-----W----Z

```
 

Know the current code is 

```
    public void addNextSibling(DetailAST ast) {
        clearBranchTokenTypes();
        clearChildCountCache(parent);
        if (ast != null) {
            // parent is set in setNextSibling
            final DetailAstImpl sibling = nextSibling;
            final DetailAstImpl astImpl = (DetailAstImpl) ast;

            if (sibling != null) {
                astImpl.setNextSibling(sibling);
                sibling.previousSibling = astImpl;
            }

            astImpl.previousSibling = this;
            setNextSibling(astImpl);
        }
    }
```

know to add the  `W` node as the sibling of `Y` in the existing structure. 
we call the method of object `Y` addNextSibling something like `Y.addNextSibling(W)` 

Then the sibling is assigned to node nextSibiling Which will be our Node `Z`
and 
astImpl will be our Node `W` 
```
 final DetailAstImpl sibling = nextSibling;
 final DetailAstImpl astImpl = (DetailAstImpl) ast;
```

Know in `if` statement  we are calling setNextSibiling where astImpl is our last means object `W` and sibling is object `Z`
```
astImpl.setNextSibling(sibling);
```

know we directly move to `setNextSibling`
```
    public void setNextSibling(DetailAST nextSibling) {
        clearBranchTokenTypes();
        clearChildCountCache(parent);
        this.nextSibling = (DetailAstImpl) nextSibling;
        if (nextSibling != null && parent != null) {
            ((DetailAstImpl) nextSibling).setParent(parent);
        }
        if (nextSibling != null) {
            ((DetailAstImpl) nextSibling).previousSibling = this;
        }
    }
```
  
here first `this.nextSibling = (DetailAstImpl) nextSibling;` here `this` keyword will point the object `W`. (The reason is https://docs.oracle.com/javase/tutorial/java/javaOO/thiskey.html#:~:text=this%20is%20a%20reference%20to%20the%20current%20object%20%E2%80%94%20the%20object%20whose%20method%20or%20constructor%20is%20being%20called here W 's object's method is called so this right know is W) and nextSibiling is Z.
```
W --> Z
```

know if we look directly at 
```
((DetailAstImpl) nextSibling).previousSibling = this;
```
here our nextSibiling is Z and this will refer to `W` means know 
```
W <----> Z 
```

know in the addNextsibiling the mutated one line `sibling.previousSibling = astImpl;` here if we recall sibling is `Z` and astIMpl is `W` 
know we again doing the stuff 
```
W <----  Z 
```

which is already done in setNextSibiling

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:150]: @rdiachenko  I hope I make the doubt clear :)

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java (hunk 1 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:143]: Yes, you are right. I am really sorry, I don't know how I have missed this. 

This may lead to major breaks in future

I will update in next push

File: src/test/java/com/puppycrawl/tools/checkstyle/DetailAstImplTest.java (hunk 3 above)
Comment (Kevin222004) [reply src/test/java/com/puppycrawl/tools/checkstyle/DetailAstImplTest.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java (hunk 1 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java]: done 

```
if (ast != null)
```

logic retrieves

**Comment by rdiachenko** [comment]
@Kevin222004 could you check failed pitest please?

**Comment by Kevin222004** [comment]
@rdiachenko Yes I am looking on it.

**Comment by romani** [comment]
spelling failure is known issue, nothing we can do for now, waiting for Sourceforge to respond

**Review by romani — APPROVED** [review]
ok to merge.

the only mixed filling I have on growing pure unit tests. But be for AST classes it is ok.

**Review by rdiachenko — APPROVED** [review]
lgtm

**Review by Vyom-Yadav — APPROVED** [review]
~LGTM!~

@Kevin222004 Please do regression testing with changes of both this and https://github.com/checkstyle/checkstyle/pull/13649

**Comment by romani** [comment]
Rebased by GitHub to be ready to report generation

**Comment by Kevin222004** [comment]
@romani @Vyom-Yadav I have updated the description section with newly generated report which has no diff

**Comment by romani** [comment]
Please fix typo

**Comment by romani** [comment]
All green, merging ...
